Add PathBuf to clap_generate::generate_to

dalance opened the pull request:
Closes #2422 

This PR adds `Result<PathBuf>` to `clap_generate::generate_to` as return value.
By this change, users can know the actual path of the generated completion file, and handle IO error instead of panic.

<!--
If your PR closes some issues, please write `Closes #XXXX`
where `XXXX` is the number of the issue you want to fix.
Each issue goes on its own line.
-->

ldm0 commented on clap_generate/src/lib.rs:
```diff
@@ -127,7 +127,11 @@ pub use shell::Shell;
+    app: &mut clap::App,
+    bin_name: S,
+    out_dir: T,
+) -> Result<PathBuf, std::io::Error>
```
As the function's usage is changed, you can update the sample code above.

pksunkara replied:
Let's do `use std::io::Error;` at the beginning of the file.

ldm0 reviewed (changes requested):
Some suggestions:

dalance commented:
@ldm0 @pksunkara Thank you for your review. I refined the usage and added `use std::io::Error`.
Additionally, the current implementation seems to be broken.
`G::file_name(app.get_bin_name().unwrap());` will panic because bin_name is not set yet.
So I added `set_app_name` first, and extracted the common part of `generate` and `generate_to` to `_generate`.

pksunkara reviewed (changes requested):
Looks good. Just one minor thing and I can merge.

dalance commented on clap_generate/src/lib.rs:
```diff
@@ -95,29 +96,39 @@ pub use shell::Shell;
+/// clap = "*"
+///
 /// [build-dependencies]
 /// clap = "*"
```
This is not required for `build.rs`. But `main.rs` requires it.
If this doesn't exist, the following error occurs.

```
$ cargo build
   Compiling test43 v0.1.0 (.../test43)
warning: completion file is generated: ".../test43/target/debug/build/test43-9a87e00ec9d60e25/out/myapp.bash"
error[E0432]: unresolved import `clap`
 --> src/cli.rs:1:5
  |
1 | use clap::{App, Arg};
  |     ^^^^ use of undeclared crate or module `clap`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0432`.
The following warnings were emitted during compilation:

warning: completion file is generated: ".../test43/target/debug/build/test43-9a87e00ec9d60e25/out/myapp.bash"

error: could not compile `test43`

To learn more, run the command again with --verbose.
```
